datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -1121,16 +1151,138 @@ impl SMJStream {
                         .collect::<Vec<_>>()
                 };
 
+            let streamed_columns_length = streamed_columns.len();
+            let buffered_columns_length = buffered_columns.len();
+
+            // Prepare the columns we apply join filter on later.
+            // Only for joined rows between streamed and buffered.
+            let filter_columns = if chunk.buffered_batch_idx.is_some() {
+                if matches!(self.join_type, JoinType::Right) {
+                    get_filter_column(&self.filter, &buffered_columns, &streamed_columns)
+                } else {
+                    get_filter_column(&self.filter, &streamed_columns, &buffered_columns)
+                }
+            } else {
+                vec![]
+            };
+
             let columns = if matches!(self.join_type, JoinType::Right) {
-                buffered_columns.extend(streamed_columns);
+                buffered_columns.extend(streamed_columns.clone());
                 buffered_columns
             } else {
                 streamed_columns.extend(buffered_columns);
                 streamed_columns
             };
 
-            self.output_record_batches
-                .push(RecordBatch::try_new(self.schema.clone(), columns)?);
+            let output_batch =
+                RecordBatch::try_new(self.schema.clone(), columns.clone())?;
+
+            // Apply join filter if any
+            if !filter_columns.is_empty() {
```

Comment:
I don't undersand why there is the check for filer columns *and* if `self.filter` is Some. I expected the check to simply be if `self.filter` is some (and the `else` case is the same for both below)

If the filter has no columns, it seems like the `else` clause does the same thing in both cases.

Thus, I wonder if we could remove the check for `filter_columns` entirely :thinking:

Reply:
Because if this joined batch is between streamed batch and null (i.e., outer joins), we don't need to handle join filter (although join filter is `Some`).

Reply:
I will add a short comment here.

Reply:
Ah, I see -- this would be for batches that don't have any matches from equality predicates anyways - which makes sense